File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;

class Solution {
    
    public boolean isBanned(String userId, String bannedId){
        if(userId.length()!=bannedId.length()) return false;
        for(int i=0; i<bannedId.length(); i++){
            if(bannedId.charAt(i)!='*'){
                if(bannedId.charAt(i)!=userId.charAt(i)) return false;
            }
        }
        return true;
    }
    
    public void dfs(int idx, String[] userIds, String[] bannedIds){
        if(idx==bannedIds.length){
            List<String> list = new ArrayList<>(bannedUsers);
            Collections.sort(list);
            String key = String.join(",",list);
            results.add(key);
            return;
        }else{
            for(int i=0; i<userIds.length; i++){
                if(!isChecked[i] && isBanned(userIds[i], bannedIds[idx])){
                    isChecked[i]=true;
                    bannedUsers.offerLast(userIds[i]);
                    dfs(idx+1, userIds, bannedIds);
                    isChecked[i]=false;
                    bannedUsers.pollLast();
                }
            }
        }
    }
    
    static boolean[] isChecked;
    static Set<String> results = new HashSet<>();
    static Deque<String> bannedUsers = new ArrayDeque<>();
    
    public int solution(String[] userIds, String[] bannedIds) {
        isChecked = new boolean[userIds.length];
        
        dfs(0, userIds, bannedIds);
        
        return results.size();
    }
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.js
function solution(userIds, bannedIds) {
    let isChecked = Array(userIds.length).fill(0);
    let result = new Set();
    
    function isBanned(userId, bannedId){
        if(userId.length!==bannedId.length) return false;
        for(let i=0; i<userId.length; i++){
            if(bannedId[i]==='*') continue;
            if(bannedId[i]!==userId[i]) return false;
        }
        return true;
    }
    
    function dfs(idx){
        if(idx==bannedIds.length){
            result.add(isChecked.join(""));
            return;
        }
        for(let i=0; i<userIds.length; i++){
            if(!isChecked[i] && isBanned(userIds[i], bannedIds[idx])){
                isChecked[i]=1;
                dfs(idx+1);
                isChecked[i]=0;
            }
        }
    }
    
    dfs(0);
    
    return result.size;
}